### PR TITLE
DEX-365 Make GET AssetGroupSighting and Sighting APIs the same

### PR DESCRIPTION
**Pull request by karenc**
- Fix integration test exception handler attribute error

  When there's a syntax error, this is shown instead of the syntax error:
  
  ```
  INTERNALERROR>   File "/home/karen/src/wildme/houston/integration_tests/conftest.py", line 265, in pytest_exception_interact
  INTERNALERROR>     if 'browser' in node.funcargs and 'session' not in node.funcargs:
  INTERNALERROR> AttributeError: 'Module' object has no attribute 'funcargs'
  ```

- Fix integration tests utils add_site_species adding duplicates

  `add_site_species` was adding a species to EDM even when it already
  existed.  The reason is because it was checking:
  
  ```
  {'commonNames': ['Example'],
   'scientificName': 'Exempli gratia',
   'id': 'fd4533c1-ce43-4826-a841-205d3f62098f'}
  == {'commonNames': ['Example'], 'scientificName': 'Exempli gratia'}
  ```
  
  which was failing because of the extra `id` field.

- Make GET AssetGroupSighting and Sighting APIs the same

  Change `AssetGroupSightingAsSightingSchema` to inherit from
  `AugmentedEdmSightingSchema` so the schema is the same as the GET
  sighting API.
  
  Add `AssetGroupSightingEncounterSchema` which is used for `encounters`
  within `AssetGroupSightingAsSightingSchema` and has the same fields as
  the GET sighting API.
  
  Add custom json encoder for decimal fields (for encounters latitude and
  longitude fields) in `config.codex.py`.

---

After this PR, the only difference between the GET sighting (red -) and asset group sighting (green +) APIs:

```diff
--- sighting_get.json   2021-11-15 15:32:57.022457152 +0000
+++ asset_group_sighting_get.json       2021-11-15 15:34:56.647491746 +0000
@@ -34,18 +34,13 @@                        
     }                                     
   ],                                      
   "comments": "None",                     
-  "createdEDM": "2021-11-15 15:32:56",    
+  "completion": 10,                       
+  "createdEDM": null,                     
   "createdHouston": "2021-11-15T15:32:56.897991+00:00",
   "customFields": {},                     
   "decimalLatitude": 63.142385,           
   "decimalLongitude": -21.596914,         
-  "encounterCounts": {                    
-    "individuals": 0,                     
-    "lifeStage": {},                      
-    "sex": {
-      "male": 1
-    }
-  },
+  "encounterCounts": {},
   "encounters": [
     {
       "country": "TEST",
@@ -69,24 +64,22 @@                        
       "sex": "male",                      
       "submitter": null,                  
       "taxonomy": {                       
-        "commonNames": [                  
-          "Example"                       
-        ],                                
-        "id": "c6439356-b02d-4edf-8059-d5d6fe1067bf",
-        "scientificName": "Exempli gratia"
+        "id": "c6439356-b02d-4edf-8059-d5d6fe1067bf"
       },                                  
       "time": "2021-11-15T15:32:41.227Z", 
       "updatedHouston": "2021-11-15T15:32:56.932337+00:00",
-      "version": 1636990376865            
+      "version": null                     
     }                                     
   ],                                      
-  "featuredAssetGuid": "d514f638-bd51-4826-a50a-94f973ccb62e",
+  "featuredAssetGuid": null,              
   "guid": "a26d9133-973d-41ec-9a4e-948700d914a4",
   "hasEdit": true,                        
   "hasView": true,                        
   "id": "a26d9133-973d-41ec-9a4e-948700d914a4",
   "locationId": "PYTEST",                 
+  "stage": "curation",                    
   "startTime": "2000-01-01T01:01:01Z",    
   "updatedHouston": "2021-11-15T15:32:56.898012+00:00",
-  "version": 1636990376865                
-}                                         
\ No newline at end of file                
+  "verbatimLocality": "",                 
+  "version": null                         
+}                                         
```

### CODE REVIEW

File: app/modules/asset_groups/models.py [hunk 1]
```diff
@@ -305,6 +305,22 @@ def getter(self):
...
+        encounters = self.config and self.config.get('encounters') or []
+        for encounter in encounters:
+            encounter['owner'] = {
+                'full_name': current_user.full_name,
+                'guid': current_user.guid,
+                'profile_fileupload': None,
+            }
+            encounter['createdHouston'] = self.created
+            encounter['updatedHouston'] = self.updated
+        log.warning(f'get_encounters {encounters}')
```
Comment (hwindsor) [inline app/modules/asset_groups/models.py:321]: warning seems a bit severe, debug seems more appropriate

Comment (karenc) [reply app/modules/asset_groups/models.py:321]: Sorry I just forgot to remove debug messages :joy:

File: app/modules/asset_groups/schemas.py [hunk 2]
```diff
@@ -42,7 +44,11 @@ class DetailedAssetGroupSightingSchema(BaseAssetGroupSightingSchema):
     Detailed Asset_group_sighting schema exposes all useful fields.
     """
 
-    assets = base_fields.Function(AssetGroupSighting.get_assets)
+    assets = base_fields.Nested(
```
Comment (hwindsor) [inline app/modules/asset_groups/schemas.py:47]: Yes that's a much cleaner way to do it. Thanks.

File: app/extensions/__init__.py [hunk 3]
```diff
@@ -193,7 +194,7 @@ class JSON(db.TypeDecorator):
     def process_bind_param(self, value, dialect):
         # Adapted from sqlalchemy/sql/sqltypes.py JSON.bind_processor
         def json_serializer(*args, **kwargs):
-            return json.dumps(*args, **kwargs, cls=flask.json.JSONEncoder)
+            return json.dumps(*args, **kwargs, cls=CustomJSONEncoder)
```
Comment (hwindsor) [inline app/extensions/__init__.py:197]: Curious as to why you made this change.

**Review by hwindsor — APPROVED** [review]
Few minor questions, good to go in I'm sure

File: app/extensions/__init__.py (hunk 3 above)
Comment (karenc) [reply app/extensions/__init__.py:197]: I added the decimal fields for AssetGroupSighting encounter latitude and longitude... but I think we might be using float elsewhere... I don't know if it matters :thinking:

Comment (karenc) [reply app/extensions/__init__.py:197]: (decimal fields are not dealt with in `flask.json.JSONEncoder`)

File: app/modules/asset_groups/models.py (hunk 1 above)
Comment (hwindsor) [reply app/modules/asset_groups/models.py:321]: Lol, I did wonder :)

File: app/extensions/__init__.py (hunk 3 above)
Comment (hwindsor) [reply app/extensions/__init__.py:197]: Wow, that is a surprise to have anything cope with floats but not integers, bu yeah I understand the change.

Comment (karenc) [reply app/extensions/__init__.py:197]: No it's not integers, it's https://docs.python.org/3/library/decimal.html, so it's like float but doesn't lose precision

**Comment by karenc** [comment]
Fixes:

```diff
diff --git a/app/extensions/__init__.py b/app/extensions/__init__.py
index 43e6eebf..c062360c 100644
--- a/app/extensions/__init__.py
+++ b/app/extensions/__init__.py
@@ -51,7 +51,7 @@ from marshmallow import Schema, validates_schema, ValidationError  # NOQA
 
 marshmallow = Marshmallow()
 
-from config import CustomJSONEncoder  # NOQA
+from config_utils import CustomJSONEncoder  # NOQA
 
 from .auth import OAuth2Provider  # NOQA
 
diff --git a/app/modules/asset_groups/models.py b/app/modules/asset_groups/models.py
index 4800f65b..8b7d2a52 100644
--- a/app/modules/asset_groups/models.py
+++ b/app/modules/asset_groups/models.py
@@ -318,7 +318,6 @@ class AssetGroupSighting(db.Model, HoustonModel):
             }
             encounter['createdHouston'] = self.created
             encounter['updatedHouston'] = self.updated
-        log.warning(f'get_encounters {encounters}')
         return encounters
 
     @classmethod
diff --git a/app/modules/asset_groups/schemas.py b/app/modules/asset_groups/schemas.py
index ead3971e..04caac9e 100644
--- a/app/modules/asset_groups/schemas.py
+++ b/app/modules/asset_groups/schemas.py
@@ -6,7 +6,7 @@ Serialization schemas for Asset_groups resources RESTful API
 
 from flask_marshmallow import base_fields
 from flask_restx_patched import ModelSchema
-from app.modules.assets.schemas import DetailedAssetSchema
+from app.modules.assets.schemas import DetailedAssetGroupAssetSchema
 from app.modules.sightings.schemas import AugmentedEdmSightingSchema
 
 from .models import AssetGroup, AssetGroupSighting
@@ -45,7 +45,7 @@ class DetailedAssetGroupSightingSchema(BaseAssetGroupSightingSchema):
     """
 
     assets = base_fields.Nested(
-        DetailedAssetSchema,
+        DetailedAssetGroupAssetSchema,
         attribute='get_assets',
         many=True,
     )
diff --git a/config.codex.py b/config.codex.py
index a95aa677..2c263663 100644
--- a/config.codex.py
+++ b/config.codex.py
@@ -1,7 +1,6 @@
 # -*- coding: utf-8 -*-
 # pylint: disable=too-few-public-methods,invalid-name,missing-docstring
 import datetime
-import decimal
 import logging
 import os
 from pathlib import Path
@@ -11,7 +10,7 @@ import string
 import pytz
 from dotenv import load_dotenv
 
-import flask
+import config_utils
 
 log = logging.getLogger(__name__)
 
@@ -29,13 +28,6 @@ PROJECT_ROOT = str(HERE)
 DATA_ROOT = Path(os.getenv('DATA_ROOT', HERE / '_db'))
 
 
-class CustomJSONEncoder(flask.json.JSONEncoder):
-    def default(self, o):
-        if isinstance(o, decimal.Decimal):
-            return str(o)
-        return super().default(o)
-
-
 class BaseConfig(object):
     # Override Flask's SERVER_NAME (default: None)
     SERVER_NAME = os.getenv('SERVER_NAME')
@@ -245,7 +237,7 @@ class BaseConfig(object):
     TIMEZONE = pytz.timezone('UTC')
 
     RESTX_JSON = {
-        'cls': CustomJSONEncoder,
+        'cls': config_utils.CustomJSONEncoder,
     }
 
     OAUTH_USER = {
diff --git a/config.mws.py b/config.mws.py
index 9d62c9b3..3f45b66f 100644
--- a/config.mws.py
+++ b/config.mws.py
@@ -10,7 +10,7 @@ import string
 import pytz
 from dotenv import load_dotenv
 
-import flask
+import config_utils
 
 log = logging.getLogger(__name__)
 
@@ -209,7 +209,7 @@ class BaseConfig(object):
     TIMEZONE = pytz.timezone('UTC')
 
     RESTX_JSON = {
-        'cls': flask.json.JSONEncoder,
+        'cls': config_utils.CustomJSONEncoder,
     }
 
     OAUTH_USER = {
diff --git a/config_utils.py b/config_utils.py
new file mode 100644
index 00000000..ef7f4bc4
--- /dev/null
+++ b/config_utils.py
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+import decimal
+
+import flask.json
+
+
+class CustomJSONEncoder(flask.json.JSONEncoder):
+    def default(self, o):
+        if isinstance(o, decimal.Decimal):
+            return str(o)
+        return super().default(o)
```

**Comment by brmscheiner** [comment]
@karenc this looks much better!! will the taxonomy differences be permanent?

File: app/extensions/__init__.py (hunk 3 above)
Comment (karenc) [reply app/extensions/__init__.py:197]: Oh I just talked to @naknomum and he said java (edm) uses double, which has the same precision as python float, so we'll just use float.

**Comment by karenc** [comment]
@brmscheiner The taxonomy returned is exactly what was sent so if we send `{"commonNames": ["Example"], "scientificName": "Exempli gratia", "id": "..."}` then we'd get that back.

Also I saw your comment on jira:

> decimalLongitude and decimalLatitude are sent as strings, they should be numbers (on the assetGroupSighting and the encounter inside the assetGroupSighting)

I'm wondering about this...  If we treat them as strings then we wouldn't lose any precision.  I'm worried that with python float / java double / javascript double, we may lose the exact value the user gives us.  Is this a valid concern?

**Comment by karenc** [comment]
^ I typed too quickly...

According to this person https://stackoverflow.com/questions/1196415/what-datatype-to-use-when-storing-latitude-and-longitude-data-in-sql-databases

> For latitudes use: `Decimal(8,6)`, and longitudes use: `Decimal(9,6)`
>
> If you're not used to precision and scale parameters, here's a format string visual:
>
> Latitude and Longitude `##.######` and `###.######`
>
> To 6 decimal places should get you to around ~10cm of accuracy on a coordinate.

I guess using python float / java double / javascript double is ok ... :sweat_smile: phew

**Comment by brmscheiner** [comment]
@karenc perfect, i can make sure to send taxonomy the same way when i post both. (or will i ever post sightings again?)

re: gps types, it looks to me like the problem is not just with assetGroupSightings but sightings as well. IIRC the decimalLatitude and decimalLongitude were numbers on the parent object, but within the encounters array they were strings

**Comment by karenc** [comment]
@brmscheiner Yeah I noticed that (latitude and longitude being string in encounters) as well.  I've fixed the asset group sighting API.  It's still a problem in the sightings API.  I'll fix it.

**Comment by brmscheiner** [comment]
<3

**Comment by karenc** [comment]
> (or will i ever post sightings again?)

@brmscheiner No, we're planning to remove `POST /api/v1/sightings/`.  We're only going to `POST /api/v1/asset_groups/` (I used this for both [integration_tests/test_sightings.py](https://github.com/WildMeOrg/houston/blob/main/integration_tests/test_sightings.py) and [integration_tests/test_asset_group_sightings.py](https://github.com/WildMeOrg/houston/blob/main/integration_tests/test_asset_group_sightings.py))

**Comment by karenc** [comment]
@hwindsor I added a commit "Change PATCH sightings API to return sighting" if you're interested :eyes:

**Review by colinwkingen — APPROVED** [review]
LGTM

**Pull request merged**